internal/controllers/machineset/machineset_controller.go

```diff
@@ -969,16 +971,37 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
 
 func (r *Reconciler) reconcileUnhealthyMachines(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, filteredMachines []*clusterv1.Machine) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
+
+	// Calculate how many in flight machines we should remediate.
+	// By default, we allow all machines to be in flight, unless otherwise.
+	maxInFlight := len(filteredMachines)
+	if strategy := ms.Spec.RemediationStrategy; strategy != nil {
+		var err error
+		maxInFlight, err = intstr.GetScaledValueFromIntOrPercent(strategy.MaxInFlight, int(ptr.Deref(ms.Spec.Replicas, 1)), true)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to calculate maxInFlight to remediate machines: %v", err)
+		}
+	}
+
 	// List all unhealthy machines.
 	machinesToRemediate := make([]*clusterv1.Machine, 0, len(filteredMachines))
 	for _, m := range filteredMachines {
```

Comment:
I think we should create a sorted copy of filtered machines to make the remediation order predictable (e.g. move the ones remediated first up to the list)